rsa.go

```diff
@@ -384,14 +384,28 @@ func newRSAKey3(isPriv bool, N, E, D, P, Q, Dp, Dq, Qinv BigInt) (C.GO_EVP_PKEY_
 		return nil, newOpenSSLError("OSSL_PARAM_BLD_new")
 	}
 	defer C.go_openssl_OSSL_PARAM_BLD_free(bld)
-	var comps = [...]struct {
+
+	type bigIntParam struct{
 		name *C.char
 		num  BigInt
-	}{
+	}
+
+	required := [...]bigIntParam{
 		{paramRSA_N, N}, {paramRSA_E, E}, {paramRSA_D, D},
-		{paramRSA_P, P}, {paramRSA_Q, Q},
-		{paramRSA_Dp, Dp}, {paramRSA_Dq, Dq}, {paramRSA_Qinv, Qinv},
 	}
+	comps := required[:]
+
+	// OpenSSL 3.0 and 3.1 required all the precomputed values if
+	// P and Q are present. See:
+	// https://github.com/openssl/openssl/pull/22334
+	if P != nil && Q != nil && Dp != nil && Dq != nil && Qinv != nil {
+		precomputed := [...]bigIntParam{
+			{paramRSA_P, P}, {paramRSA_Q, Q},
+			{paramRSA_Dp, Dp}, {paramRSA_Dq, Dq}, {paramRSA_Qinv, Qinv},
+		}
+		comps = append(comps, precomputed[:]...)
```

Comment:
This will have to grow the slice, as `comps` length is `3` at this point, producing an allocation. Do something like this when initializing `comps` to avoid the allocation:

```go
comps := make([]bigIntParam, 0, 8)
comps = append(comps, required...)
```

Reply:
Thanks; rewritten in that way.